consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/TestContext.java

```diff
@@ -0,0 +1,290 @@
+import tech.pegasys.pantheon.ethereum.core.BlockBody;
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.core.BlockHeaderTestFixture;
+import tech.pegasys.pantheon.ethereum.core.Hash;
+import tech.pegasys.pantheon.ethereum.core.MiningParameters;
+import tech.pegasys.pantheon.ethereum.core.PendingTransactions;
+import tech.pegasys.pantheon.ethereum.core.Util;
+import tech.pegasys.pantheon.ethereum.core.Wei;
+import tech.pegasys.pantheon.ethereum.db.WorldStateArchive;
+import tech.pegasys.pantheon.ethereum.mainnet.BlockHeaderValidator;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.util.bytes.BytesValue;
+import tech.pegasys.pantheon.util.uint.UInt256;
+
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.Iterables;
+
+/*
+Responsible for creating an environment in which integration testing can be conducted.
+
+The test setup is a 'n' node network, the first of which can be considered the "local" node.
+All allValidators are in the genesis block's extraData. Validator
+
+ */
+public class TestContext {
+
+  public static final int EPOCH_LENGTH = 10_000;
+  public static final int BLOCK_TIMER_SEC = 3;
+  public static final int ROUND_TIMER_SEC = 12;
+
+  private Map<Address, ValidatorPeer> remotePeers;
+  private final MutableBlockchain blockchain;
+  private final IbftController controller;
+  private final IbftFinalState finalState;
+
+  public TestContext(
+      final Map<Address, ValidatorPeer> remotePeers,
+      final MutableBlockchain blockchain,
+      final IbftController controller,
+      final IbftFinalState finalState) {
+    this.remotePeers = remotePeers;
+    this.blockchain = blockchain;
+    this.controller = controller;
+    this.finalState = finalState;
+  }
+
+  public Collection<ValidatorPeer> getRemotePeers() {
+    return remotePeers.values();
+  }
+
+  public MutableBlockchain getBlockchain() {
+    return blockchain;
+  }
+
+  public IbftController getController() {
+    return controller;
+  }
+
+  public MessageFactory getLocalNodeMessageFactory() {
+    return finalState.getMessageFactory();
+  }
+
+  public Block createBlockForProposal(final int round, final long timestamp) {
+    return finalState
+        .getBlockCreatorFactory()
+        .create(blockchain.getChainHeadHeader(), round)
+        .createBlock(timestamp);
+  }
+
+  public RoundSpecificNodeRoles getRoundSpecificRoles(final ConsensusRoundIdentifier roundId) {
+    // This will return NULL if the LOCAL node is the proposer for the specified round
+    final Address proposerAddress = finalState.getProposerForRound(roundId);
```

Comment:
Would really rather not - otherwise, everytime you go to use "getProposer()" you then have to perform a "get()".
ATM: The test script writer is expected to know how their network is laid out, and thus should appreciate that the "remote nodes" do not have a proposer amongst them....

Reply:
...and there's always the NPE.

Reply:
yeah, I kinda figured the NPE failing the test would result in someone checking out what's gone wrong ... ultimately is a badly formed test so should fail.